File: dotnet/PolymarketBot/Services/MarketScanner.cs
using System.Text.Json;
using Microsoft.Extensions.Logging;
using PolymarketBot.Models;

namespace PolymarketBot.Services;

public sealed class MarketScanner
{
    private static readonly Dictionary<string, string[]> CategoryKeywords = new()
    {
        ["politics"] = ["president", "election", "congress", "senate", "governor", "vote", "party",
                        "democrat", "republican", "trump", "biden", "political", "inaugur",
                        "legislation", "supreme court", "cabinet", "impeach", "primary"],
        ["geopolitics"] = ["iran", "israel", "strike", "invade", "invasion", "war", "military",
                           "nato", "sanction", "nuclear", "missile", "ceasefire", "peace deal",
                           "china", "taiwan", "russia", "ukraine", "north korea", "tariff"],
        ["sports"] = ["nfl", "nba", "mlb", "nhl", "soccer", "football", "basketball", "baseball",
                      "tennis", "ufc", "fight", "championship", "super bowl", "world series",
                      "premier league", "match", "game", "serie a", "ncaa", "ligue 1",
                      "olympics", "medal", "la liga", "bundesliga", "win on 202", "rio open",
                      "open:", "grand slam"],
        ["crypto"] = ["bitcoin", "btc", "ethereum", "eth", "crypto", "solana", "sol", "token",
                      "defi", "blockchain", "coin", "memecoin", "fdv", "airdrop"],
        ["tech"] = ["ai model", "claude", "gpt", "openai", "anthropic", "google ai", "apple",
                    "microsoft", "tesla", "spacex", "launch", "release", "chip", "semiconductor"],
        ["social_media"] = ["tweet", "post", "elon musk", "follower", "subscriber", "tiktok",
                            "youtube", "instagram", "x.com"],
        ["weather"] = ["weather", "temperature", "hurricane", "storm", "rainfall", "snow", "climate"],
        ["entertainment"] = ["oscar", "grammy", "emmy", "movie", "film", "tv", "show", "album",
                             "music", "celebrity", "award", "box office"],
        ["finance"] = ["fed", "interest rate", "inflation", "gdp", "stock", "market", "s&p",
                       "nasdaq", "dow", "recession", "unemployment", "spx", "treasury"],
    };

    private readonly BotConfig _config;
    private readonly HttpClient _http;
    private readonly ILogger<MarketScanner> _log;
    private readonly string _baseUrl;
    private readonly string _clobHost;

    public MarketScanner(BotConfig config, HttpClient http, ILogger<MarketScanner> log)
    {
        _config = config;
        _http = http;
        _log = log;
        _baseUrl = config.GammaApiHost;
        _clobHost = config.ClobHost;
    }

    public async Task<List<MarketInfo>> ScanAsync(CancellationToken ct = default)
    {
        var rawEvents = await FetchAllEventsAsync(ct);
        var markets = new List<MarketInfo>();

        foreach (var evt in rawEvents)
        {
            var eventTitle = evt.GetPropertyOrDefault("title", "");
            var eventSlug = evt.GetPropertyOrDefault("slug", "");
            var description = evt.GetPropertyOrDefault("description", "");
            var category = Categorize(eventTitle, eventSlug);

            if (evt.TryGetProperty("markets", out var marketsEl) && marketsEl.ValueKind == JsonValueKind.Array)
            {
                foreach (var mkt in marketsEl.EnumerateArray())
                {
                    var parsed = ParseMarket(mkt, eventTitle, description, category);
                    if (parsed is not null)
                        markets.Add(parsed);
                }
            }
        }

        markets.Sort((a, b) => b.Volume24Hr.CompareTo(a.Volume24Hr));
        _log.LogInformation("Scan complete: {EventCount} events, {MarketCount} eligible markets",
            rawEvents.Count, markets.Count);
        return markets;
    }

    private async Task<List<JsonElement>> FetchAllEventsAsync(CancellationToken ct)
    {
        var all = new List<JsonElement>();
        var offset = 0;
        const int limit = 100;

        while (true)
        {
            var page = await FetchEventsPageAsync(offset, limit, ct);
            if (page.Count == 0) break;
            all.AddRange(page);
            if (page.Count < limit) break;
            offset += limit;
        }

        return all;
    }

    private async Task<List<JsonElement>> FetchEventsPageAsync(int offset, int limit, CancellationToken ct)
    {
        var url = $"{_baseUrl}/events?active=true&closed=false&limit={limit}&offset={offset}";

        for (var attempt = 0; attempt < 3; attempt++)
        {
            try
            {
                var resp = await _http.GetAsync(url, ct);

                if ((int)resp.StatusCode == 429)
                {
                    var wait = (int)Math.Pow(2, attempt);
                    _log.LogWarning("Rate limited on /events, retrying in {Wait}s", wait);
                    await Task.Delay(wait * 1000, ct);
                    continue;
                }

                resp.EnsureSuccessStatusCode();
                var json = await resp.Content.ReadAsStringAsync(ct);
                var doc = JsonDocument.Parse(json);

                if (doc.RootElement.ValueKind == JsonValueKind.Array)
                    return doc.RootElement.EnumerateArray().ToList();

                return [];
            }
            catch (Exception ex) when (ex is not OperationCanceledException)
            {
                if (attempt < 2)
                {
                    var wait = (int)Math.Pow(2, attempt);
                    _log.LogWarning("Error fetching events (attempt {Attempt}): {Error}, retrying in {Wait}s",
                        attempt + 1, ex.Message, wait);
                    await Task.Delay(wait * 1000, ct);
                }
                else
                {
                    _log.LogError("Failed to fetch events after 3 attempts: {Error}", ex.Message);
                    return [];
                }
            }
        }

        return [];
    }

    private MarketInfo? ParseMarket(JsonElement mkt, string eventTitle, string description, string category)
    {
        try
        {
            if (!mkt.GetPropertyOrDefault("active", false) || mkt.GetPropertyOrDefault("closed", false))
                return null;

            var outcomes = ParseJsonStringOrArray(mkt, "outcomes");
            if (outcomes.Count != 2) return null;

            var prices = ParseJsonStringOrArray(mkt, "outcomePrices");
            if (prices.Count != 2) return null;

            var yesPrice = double.Parse(prices[0]);
            var noPrice = double.Parse(prices[1]);

            var tokens = ParseJsonStringOrArray(mkt, "clobTokenIds");
            if (tokens.Count != 2) return null;

            var liquidity = mkt.GetDoubleOrDefault("liquidity");
            var volume = mkt.GetDoubleOrDefault("volume");
            var volume24Hr = mkt.GetDoubleOrDefault("volume24hr");

            if (liquidity < _config.MinLiquidity) return null;
            if (volume24Hr < _config.MinVolume24Hr) return null;

            // Price filter — skip markets where neither side is in the tradeable range
            // Markets at extreme prices (e.g. YES=0.001, NO=0.999) have no FOK liquidity
            var minP = _config.MinMarketPrice;
            var maxP = 1.0 - minP;
            var yesInRange = yesPrice >= minP && yesPrice <= maxP;
            var noInRange = noPrice >= minP && noPrice <= maxP;
            if (!yesInRange && !noInRange)
                return null;

            var endDateStr = mkt.GetPropertyOrDefault("endDate", "");
            if (!string.IsNullOrEmpty(endDateStr))
            {
                if (DateTimeOffset.TryParse(endDateStr, out var endDate))
                {
                    var hoursLeft = (endDate - DateTimeOffset.UtcNow).TotalHours;
                    if (hoursLeft < _config.MinTimeToResolutionHours)
                        return null;
                }
            }

            var bestBid = mkt.GetDoubleOrDefault("bestBid");
            var bestAsk = mkt.GetDoubleOrDefault("bestAsk");
            var spread = bestAsk > bestBid ? bestAsk - bestBid : 0.0;

            return new MarketInfo
            {
                ConditionId = mkt.GetPropertyOrDefault("conditionId", ""),
                Question = mkt.GetPropertyOrDefault("question", eventTitle),
                Slug = mkt.GetPropertyOrDefault("slug", ""),
                OutcomeYesPrice = yesPrice,
                OutcomeNoPrice = noPrice,
                TokenIdYes = tokens[0],
                TokenIdNo = tokens[1],
                Liquidity = liquidity,
                Volume = volume,
                Volume24Hr = volume24Hr,
                BestBid = bestBid,
                BestAsk = bestAsk,
                Spread = spread,
                EndDate = endDateStr,
                Category = category,
                EventTitle = eventTitle,
                Description = mkt.GetPropertyOrDefault("description", description),
            };
        }
        catch (Exception ex)
        {
            _log.LogDebug("Failed to parse market: {Error}", ex.Message);
            return null;
        }
    }

    private static List<string> ParseJsonStringOrArray(JsonElement parent, string propertyName)
    {
        if (!parent.TryGetProperty(propertyName, out var el))
            return [];

        if (el.ValueKind == JsonValueKind.Array)
            return el.EnumerateArray().Select(e => e.GetString() ?? "").ToList();

        if (el.ValueKind == JsonValueKind.String)
        {
            var str = el.GetString() ?? "[]";
            var arr = JsonDocument.Parse(str).RootElement;
            if (arr.ValueKind == JsonValueKind.Array)
                return arr.EnumerateArray().Select(e => e.GetString() ?? e.GetRawText().Trim('"')).ToList();
        }

        return [];
    }

    private static string Categorize(string title, string slug)
    {
        var text = $"{title} {slug}".ToLowerInvariant();
        foreach (var (category, keywords) in CategoryKeywords)
        {
            if (keywords.Any(kw => text.Contains(kw, StringComparison.OrdinalIgnoreCase)))
                return category;
        }
        return "other";
    }

    public async Task<Dictionary<string, string>?> CheckMarketResolutionAsync(
        string conditionId, CancellationToken ct = default)
    {
        try
        {
            var url = $"{_clobHost}/markets/{conditionId}";
            var resp = await _http.GetAsync(url, ct);
            if (resp.StatusCode == System.Net.HttpStatusCode.NotFound)
                return null;
            resp.EnsureSuccessStatusCode();
            var json = await resp.Content.ReadAsStringAsync(ct);
            var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            if (!root.GetPropertyOrDefault("closed", false))
                return null;

            // CLOB returns tokens array with winner flag
            if (root.TryGetProperty("tokens", out var tokensEl) && tokensEl.ValueKind == JsonValueKind.Array)
            {
                foreach (var token in tokensEl.EnumerateArray())
                {
                    if (token.GetPropertyOrDefault("winner", false))
                    {
                        var outcome = token.GetPropertyOrDefault("outcome", "").ToUpperInvariant();
                        if (outcome == "YES")
                            return new Dictionary<string, string> { ["winning_side"] = "YES" };
                        if (outcome == "NO")
                            return new Dictionary<string, string> { ["winning_side"] = "NO" };
                    }
                }
            }

            return null;
        }
        catch (Exception ex) when (ex is not OperationCanceledException)
        {
            _log.LogDebug("Resolution check failed for {ConditionId}: {Error}",
                conditionId[..Math.Min(conditionId.Length, 20)], ex.Message);
            return null;
        }
    }

    public async Task<Dictionary<string, double>> GetMarketPricesAsync(
        IEnumerable<string> tokenIds, CancellationToken ct = default)
    {
        var prices = new Dictionary<string, double>();
        foreach (var tid in tokenIds)
        {
            var p = await GetMarketPriceAsync(tid, ct);
            if (p.HasValue && p.Value > 0)
                prices[tid] = p.Value;
        }
        return prices;
    }

    public async Task<double?> GetMarketPriceAsync(string tokenId, CancellationToken ct = default)
    {
        try
        {
            var url = $"{_clobHost}/midpoint?token_id={tokenId}";
            var resp = await _http.GetAsync(url, ct);
            resp.EnsureSuccessStatusCode();
            var json = await resp.Content.ReadAsStringAsync(ct);
            var doc = JsonDocument.Parse(json);
            if (doc.RootElement.TryGetProperty("mid", out var mid))
            {
                if (mid.ValueKind == JsonValueKind.Number)
                    return mid.GetDouble();
                if (mid.ValueKind == JsonValueKind.String && double.TryParse(mid.GetString(), out var val))
                    return val;
            }
            return null;
        }
        catch
        {
            return null;
        }
    }
}

internal static class JsonElementExtensions
{
    public static string GetPropertyOrDefault(this JsonElement el, string name, string defaultValue)
    {
        if (el.TryGetProperty(name, out var prop) && prop.ValueKind == JsonValueKind.String)
            return prop.GetString() ?? defaultValue;
        return defaultValue;
    }

    public static bool GetPropertyOrDefault(this JsonElement el, string name, bool defaultValue)
    {
        if (el.TryGetProperty(name, out var prop))
        {
            if (prop.ValueKind == JsonValueKind.True) return true;
            if (prop.ValueKind == JsonValueKind.False) return false;
        }
        return defaultValue;
    }

    public static double GetDoubleOrDefault(this JsonElement el, string name, double defaultValue = 0.0)
    {
        if (!el.TryGetProperty(name, out var prop)) return defaultValue;
        if (prop.ValueKind == JsonValueKind.Number) return prop.GetDouble();
        if (prop.ValueKind == JsonValueKind.String && double.TryParse(prop.GetString(), out var val)) return val;
        if (prop.ValueKind == JsonValueKind.Null) return defaultValue;
        return defaultValue;
    }
}

File: python/market_scanner.py
"""Gamma API integration for market discovery and filtering."""

import json
import logging
import time
from datetime import datetime, timezone
from typing import Optional

import requests

from config import BotConfig
from models import MarketInfo

log = logging.getLogger("bot.scanner")

# Keywords for rough category classification from event slugs/titles
CATEGORY_KEYWORDS = {
    "politics": ["president", "election", "congress", "senate", "governor", "vote", "party",
                  "democrat", "republican", "trump", "biden", "political", "inaugur",
                  "legislation", "supreme court", "cabinet", "impeach", "primary"],
    "geopolitics": ["iran", "israel", "strike", "invade", "invasion", "war", "military",
                    "nato", "sanction", "nuclear", "missile", "ceasefire", "peace deal",
                    "china", "taiwan", "russia", "ukraine", "north korea", "tariff"],
    "sports": ["nfl", "nba", "mlb", "nhl", "soccer", "football", "basketball", "baseball",
               "tennis", "ufc", "fight", "championship", "super bowl", "world series",
               "premier league", "match", "game", "serie a", "ncaa", "ligue 1",
               "olympics", "medal", "la liga", "bundesliga", "win on 202", "rio open",
               "open:", "grand slam"],
    "crypto": ["bitcoin", "btc", "ethereum", "eth", "crypto", "solana", "sol", "token",
               "defi", "blockchain", "coin", "memecoin", "fdv", "airdrop"],
    "tech": ["ai model", "claude", "gpt", "openai", "anthropic", "google ai", "apple",
             "microsoft", "tesla", "spacex", "launch", "release", "chip", "semiconductor"],
    "social_media": ["tweet", "post", "elon musk", "follower", "subscriber", "tiktok",
                     "youtube", "instagram", "x.com"],
    "weather": ["weather", "temperature", "hurricane", "storm", "rainfall", "snow", "climate"],
    "entertainment": ["oscar", "grammy", "emmy", "movie", "film", "tv", "show", "album",
                      "music", "celebrity", "award", "box office"],
    "finance": ["fed", "interest rate", "inflation", "gdp", "stock", "market", "s&p",
                "nasdaq", "dow", "recession", "unemployment", "spx", "treasury"],
}


class MarketScanner:
    def __init__(self, config: BotConfig):
        self.config = config
        self.session = requests.Session()
        self.session.headers.update({"Accept": "application/json"})
        self.base_url = config.gamma_api_host

    def scan(self) -> list[MarketInfo]:
        """Fetch all active markets, filter, and return eligible MarketInfo list."""
        raw_events = self._fetch_all_events()
        markets = []

        for event in raw_events:
            event_title = event.get("title", "")
            event_slug = event.get("slug", "")
            description = event.get("description", "")
            category = self._categorize(event_title, event_slug)

            for mkt in event.get("markets", []):
                parsed = self._parse_market(mkt, event_title, description, category)
                if parsed is not None:
                    markets.append(parsed)

        # Sort by 24h volume descending (highest activity first)
        markets.sort(key=lambda m: m.volume_24hr, reverse=True)
        log.info(f"Scan complete: {len(raw_events)} events, {len(markets)} eligible markets")
        return markets

    def _fetch_all_events(self) -> list[dict]:
        """Fetch all active events with pagination."""
        all_events = []
        offset = 0
        limit = 100

        while True:
            page = self._fetch_events_page(offset, limit)
            if not page:
                break
            all_events.extend(page)
            if len(page) < limit:
                break
            offset += limit

        return all_events

    def _fetch_events_page(self, offset: int, limit: int = 100) -> list[dict]:
        """Single page fetch with retry logic."""
        url = f"{self.base_url}/events"
        params = {
            "active": "true",
            "closed": "false",
            "limit": limit,
            "offset": offset,
        }

        for attempt in range(3):
            try:
                resp = self.session.get(url, params=params, timeout=30)
                if resp.status_code == 429:
                    wait = 2 ** attempt
                    log.warning(f"Rate limited on /events, retrying in {wait}s")
                    time.sleep(wait)
                    continue
                resp.raise_for_status()
                return resp.json()
            except requests.RequestException as e:
                if attempt < 2:
                    wait = 2 ** attempt
                    log.warning(f"Error fetching events (attempt {attempt + 1}): {e}, retrying in {wait}s")
                    time.sleep(wait)
                else:
                    log.error(f"Failed to fetch events after 3 attempts: {e}")
                    return []

        return []

    def _parse_market(self, mkt: dict, event_title: str, description: str,
                      category: str) -> Optional[MarketInfo]:
        """Parse a single market dict into MarketInfo. Returns None if filtered out."""
        try:
            # Must be active and accepting orders
            if not mkt.get("active", False) or mkt.get("closed", False):
                return None

            # Parse outcome prices (JSON-encoded string)
            outcomes_raw = mkt.get("outcomes", "[]")
            if isinstance(outcomes_raw, str):
                outcomes = json.loads(outcomes_raw)
            else:
                outcomes = outcomes_raw

            # Only binary markets
            if len(outcomes) != 2:
                return None

            prices_raw = mkt.get("outcomePrices", "[]")
            if isinstance(prices_raw, str):
                prices = json.loads(prices_raw)
            else:
                prices = prices_raw

            if len(prices) != 2:
                return None

            yes_price = float(prices[0])
            no_price = float(prices[1])

            # Parse token IDs (JSON-encoded string)
            tokens_raw = mkt.get("clobTokenIds", "[]")
            if isinstance(tokens_raw, str):
                tokens = json.loads(tokens_raw)
            else:
                tokens = tokens_raw

            if len(tokens) != 2:
                return None

            token_yes = tokens[0]
            token_no = tokens[1]

            # Liquidity and volume filters
            liquidity = float(mkt.get("liquidity", 0) or 0)
            volume = float(mkt.get("volume", 0) or 0)
            volume_24hr = float(mkt.get("volume24hr", 0) or 0)

            if liquidity < self.config.min_liquidity:
                return None
            if volume_24hr < self.config.min_volume_24hr:
                return None

            # Price filter — skip markets where neither side is in the tradeable range
            # Markets at extreme prices (e.g. YES=0.001, NO=0.999) have no FOK liquidity
            min_p = self.config.min_market_price
            max_p = 1.0 - min_p
            if not (min_p <= yes_price <= max_p or min_p <= no_price <= max_p):
                return None

            # Time to resolution filter
            end_date_str = mkt.get("endDate", "")
            if end_date_str:
                try:
                    end_date = datetime.fromisoformat(end_date_str.replace("Z", "+00:00"))
                    now = datetime.now(timezone.utc)
                    hours_left = (end_date - now).total_seconds() / 3600
                    if hours_left < self.config.min_time_to_resolution_hours:
                        return None
                except (ValueError, TypeError):
                    pass  # If we can't parse the date, don't filter on it

            # Spread
            best_bid = float(mkt.get("bestBid", 0) or 0)
            best_ask = float(mkt.get("bestAsk", 0) or 0)
            spread = best_ask - best_bid if best_ask > best_bid else 0.0

            question = mkt.get("question", event_title)
            slug = mkt.get("slug", "")
            mkt_description = mkt.get("description", description)

            return MarketInfo(
                condition_id=mkt.get("conditionId", ""),
                question=question,
                slug=slug,
                outcome_yes_price=yes_price,
                outcome_no_price=no_price,
                token_id_yes=token_yes,
                token_id_no=token_no,
                liquidity=liquidity,
                volume=volume,
                volume_24hr=volume_24hr,
                best_bid=best_bid,
                best_ask=best_ask,
                spread=spread,
                end_date=end_date_str,
                category=category,
                event_title=event_title,
                description=mkt_description or "",
            )

        except (KeyError, ValueError, TypeError) as e:
            log.debug(f"Failed to parse market: {e}")
            return None

    def _categorize(self, title: str, slug: str) -> str:
        """Classify market category from event title/slug using keyword matching."""
        text = f"{title} {slug}".lower()
        for category, keywords in CATEGORY_KEYWORDS.items():
            if any(kw in text for kw in keywords):
                return category
        return "other"

    def check_market_resolution(self, condition_id: str) -> Optional[dict]:
        """Check if a market has resolved via CLOB API.
        Returns {"winning_side": "YES"|"NO"} if resolved, None if still active."""
        try:
            url = f"{self.config.clob_host}/markets/{condition_id}"
            resp = self.session.get(url, timeout=10)
            if resp.status_code == 404:
                return None
            resp.raise_for_status()
            data = resp.json()

            if not data.get("closed", False):
                return None

            # CLOB returns tokens array with winner flag
            tokens = data.get("tokens", [])
            for token in tokens:
                if token.get("winner", False):
                    outcome = token.get("outcome", "").upper()
                    if outcome == "YES":
                        return {"winning_side": "YES"}
                    elif outcome == "NO":
                        return {"winning_side": "NO"}

            # Market closed but no winner flagged yet
            return None
        except Exception as e:
            log.debug(f"Resolution check failed for {condition_id[:20]}...: {e}")
            return None

    def get_market_prices(self, token_ids: list[str]) -> dict[str, float]:
        """Fetch current prices for multiple tokens. Returns dict of token_id -> midpoint price."""
        prices = {}
        for tid in token_ids:
            p = self.get_market_price(tid)
            if p is not None and p > 0:
                prices[tid] = p
        return prices

    def get_market_price(self, token_id: str) -> Optional[float]:
        """Fetch current price for a single token from the CLOB API."""
        try:
            url = f"{self.config.clob_host}/midpoint"
            params = {"token_id": token_id}
            resp = self.session.get(url, params=params, timeout=10)
            resp.raise_for_status()
            data = resp.json()
            return float(data.get("mid", 0))
        except Exception as e:
            log.debug(f"Failed to get price for {token_id[:20]}...: {e}")
            return None
